### PR TITLE
BUG: Allow all int types for merge (GH28870)

### LINKED ISSUE #28870: Merge_asof() Requires specific int type, not reflected in error or documentation
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

left = pd.DataFrame({'ts_int': [0, 100, 200], 'left_val': [1, 2, 3]})
right = pd.DataFrame({'ts_int': [50, 150, 250], 'right_val': [1, 2, 3]})
left['ts_int'] = left['ts_int'].astype(int)
right['ts_int'] = right['ts_int'].astype(int)
pd.merge_asof(left, right, on='ts_int', tolerance=100)

> pandas.errors.MergeError: key must be integer, timestamp or float

print(left['ts_int'].dtype)
>>> int32
print(right['ts_int'].dtype)
>>> int32
```
#### Problem description

merge_asof() throws the MergeError `key must be integer, timestamp or float` even when the data used to merge on is a valid integer type.

This happens with all types of int with the exception of int64, as a result of the check performed when a tolerance is passed _(pandas/core/reshape/merge.py:1641)_. This check uses the `is_int64_dtype(lt)`, which will ofcourse return `False` for every type of int that isn't `int64`. This requirement of the `on` key being int64 is however not documented in the docs, nor is it reflected from the raised error.

#### Expected Output

The expected output can go either of two ways. If it should be possible to perform the merge on every integer type, the expected output would be:
```
   ts_int  left_val  right_val
0       0         1        NaN
1     100         2        1.0
2     200         3        2.0
```

If it shouldn't be possible, I'd expect the returned MergeError and/or the docs to reflect that the integer type that should be used as key to merge on has to be of int64 type.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : None.None

pandas           : 0.25.1
numpy            : 1.17.2
pytz             : 2019.2
dateutil         : 2.8.0
pip              : 19.2.3
setuptools       : 41.2.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
</details>

### PR DESCRIPTION
**Pull request by jjlkant**
- [x] closes #28870
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by mroeschke — CHANGES_REQUESTED** [review]
* Needs a test to confirm the behavior has been fixed.
* Needs a new whatsnew entry (in 0.25.2)

File: doc/source/whatsnew/v0.25.2.rst [hunk 1]
```diff
@@ -86,7 +86,7 @@ Groupby/resample/rolling
 Reshaping
 ^^^^^^^^^
 
--
+- Fix to ensure all int dtypes can be used in :meth:`pandas.core.reshape.merge_asof` when using a tolerance value (:issue:`28870`).
```
Comment (mroeschke) [inline doc/source/whatsnew/v0.25.2.rst]: `` :func:`merge_asof` `` and maybe clarify that any non-int64 integer previously raised an erroneous `MergeError`

File: pandas/tests/reshape/merge/test_merge_asof.py [hunk 2]
```diff
@@ -1287,3 +1287,19 @@ def test_timedelta_tolerance_nearest(self):
         )
 
         assert_frame_equal(result, expected)
+
+    def test_int_type_tolerance(self):
+        # GH #28870
+
+        left = pd.DataFrame({"a": [0, 100, 200], "left_val": [1, 2, 3]})
+        right = pd.DataFrame({"a": [50, 150, 250], "right_val": [1, 2, 3]})
+        left["a"] = left["a"].astype(np.int32)
```
Comment (mroeschke) [inline pandas/tests/reshape/merge/test_merge_asof.py]: Could you `pytest.mark.parameterize` over more int dtypes. e.g. `[int, "int64", "int32", "int16", "int8", "uint64", "uint32", "uint16", "uint8"]`

**Review by mroeschke — APPROVED** [review]

File: doc/source/whatsnew/v0.25.2.rst [hunk 3]
```diff
@@ -86,7 +86,7 @@ Groupby/resample/rolling
 Reshaping
 ^^^^^^^^^
 
--
+- Fix to ensure all int dtypes can be used in :func:`merge_asof` when using a tolerance value. Before every non-int64 type would raise an erroneous ``MergeError`` (:issue:`28870`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.2.rst]: moved to 1.0

Before -> Previously

File: pandas/tests/reshape/merge/test_merge_asof.py [hunk 4]
```diff
@@ -1287,3 +1287,23 @@ def test_timedelta_tolerance_nearest(self):
         )
 
         assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize(
+        "dtype",
+        [int, "int64", "int32", "int16", "int8", "uint64", "uint32", "uint16", "uint8"],
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge_asof.py]: use the any_int_dtype fixture instead

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jjlkant

**Pull request merged**